Parsing trouble when a column is called "vector"

On a schema like this
```
CREATE TABLE cassio_tutorials.my_vector_db_table (
    row_id text PRIMARY KEY,
    attributes_blob text,
    body_blob text,
    metadata_s map<text, text>,
    vector vector<float, 1536>
);
```

it happens that `dsbulk unload` gets confused by `vector` and the unload operation fails with something like

```
Operation UNLOAD_20230927-152011-882206 failed: Invalid query: 'SELECT row_id, attributes_blob, body_blob, metadata_s, vector FROM cassio_tutorials.my_vector_db_table WHERE token(row_id) > :start AND token(row_id) <= :end' could not be parsed at line 1:55: mismatched input 'vector' expecting {'(', ':', '[', '-', '{', K_AS, K_KEY, K_PER, K_PARTITION, K_DISTINCT, K_COUNT, K_VALUES, K_TIMESTAMP, K_TTL, K_CAST, K_TYPE, K_FILTERING, K_CONTAINS, K_GROUP, K_CLUSTERING, K_ASCII, K_BIGINT, K_BLOB, K_BOOLEAN, K_COUNTER, K_DECIMAL, K_DOUBLE, K_DURATION, K_FLOAT, K_INET, K_INT, K_SMALLINT, K_TINYINT, K_TEXT, K_UUID, K_VARCHAR, K_VARINT, K_TIMEUUID, K_TOKEN, K_WRITETIME, K_DATE, K_TIME, K_NULL, K_EXISTS, K_MAP, K_LIST, K_NAN, K_INFINITY, K_TUPLE, K_FROZEN, K_JSON, K_LIKE, STRING_LITERAL, QUOTED_NAME, INTEGER, '?', FLOAT, BOOLEAN, DURATION, IDENT, HEXNUMBER, UUID}.
   Caused by: InputMismatchException (no message).
```

v1.11

--- a/cql/src/main/antlr4/com/datastax/oss/dsbulk/generated/cql3/Cql.g4
+++ b/cql/src/main/antlr4/com/datastax/oss/dsbulk/generated/cql3/Cql.g4
@@ -449,7 +449,7 @@ relation
     | cident K_IN singleColumnInValues
     | cident K_CONTAINS (K_KEY)? term
     | cident '[' term ']' relationType term
-    | cident K_ANN_OF term
+    | cident K_ANN K_OF term
     | tupleOfIdentifiers
       ( K_IN
           ( '(' ')'
@@ -499,6 +499,7 @@ comparatorType
     | collectionType
     | tupleType
     | userTypeName
+    | vectorType
     | K_FROZEN '<' comparatorType '>'
     ;
 
@@ -524,7 +525,6 @@ nativeType
     | K_TIMEUUID
     | K_DATE
     | K_TIME
-    | K_VECTOR
     ;
 
 collectionType
@@ -537,6 +537,9 @@ tupleType
     : K_TUPLE '<' comparatorType (',' comparatorType)* '>'
     ;
 
+vectorType
+    : K_VECTOR '<' comparatorType ',' INTEGER '>'
+    ;
 
 // Basically the same as cident, but we need to exlude existing CQL3 types
 // (which for some reason are not reserved otherwise)
@@ -573,6 +576,8 @@ basicUnreservedKeyword
         | K_PER
         | K_PARTITION
         | K_GROUP
+        | K_VECTOR
+        | K_ANN
         )
     ;
 
@@ -660,7 +665,8 @@ K_DEFAULT:     D E F A U L T;
 K_UNSET:       U N S E T;
 K_LIKE:        L I K E;
 
-K_ANN_OF:      A N N WS+ O F;
+K_ANN:         A N N;
+K_OF:          O F;
 
 // Case-insensitive alpha characters
 fragment A: ('a'|'A');

--- a/cql/src/main/antlr4/com/datastax/oss/dsbulk/generated/cql3/Cql.g4
+++ b/cql/src/main/antlr4/com/datastax/oss/dsbulk/generated/cql3/Cql.g4
@@ -524,6 +524,7 @@ nativeType
     | K_TIMEUUID
     | K_DATE
     | K_TIME
+    | K_VECTOR
     ;
 
 collectionType

--- a/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
+++ b/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
@@ -34,8 +34,6 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
-import javax.management.Query;
-
 class QueryInspectorTest {
 
   private static final CQLWord NOW = CQLWord.fromInternal("now");

--- a/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
+++ b/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
@@ -657,6 +657,17 @@ class QueryInspectorTest {
 
   @ParameterizedTest
   @MethodSource
+  void should_detect_unsupported_vector_selector(String query, boolean expected) {
+    assertThat(new QueryInspector(query).hasUnsupportedSelectors()).isEqualTo(expected);
+  }
+
+  @SuppressWarnings("unused")
+  static List<Arguments> should_detect_unsupported_vector_selector() {
+    return Lists.newArrayList(arguments("SELECT (vector<int,3>)[0.123] FROM ks.t1", true));
+  }
+
+  @ParameterizedTest
+  @MethodSource
   void should_detect_functions_in_assignments(
       String query, int expectedTotalAssignments, FunctionCall expectedValue) {
     QueryInspector inspector = new QueryInspector(query);

--- a/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
+++ b/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
@@ -34,6 +34,8 @@ import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
+import javax.management.Query;
+
 class QueryInspectorTest {
 
   private static final CQLWord NOW = CQLWord.fromInternal("now");
@@ -655,15 +657,11 @@ class QueryInspectorTest {
         arguments("SELECT * FROM ks.t1", false));
   }
 
-  @ParameterizedTest
-  @MethodSource
-  void should_detect_unsupported_vector_selector(String query, boolean expected) {
-    assertThat(new QueryInspector(query).hasUnsupportedSelectors()).isEqualTo(expected);
-  }
-
-  @SuppressWarnings("unused")
-  static List<Arguments> should_detect_unsupported_vector_selector() {
-    return Lists.newArrayList(arguments("SELECT (vector<int,3>)[0.123] FROM ks.t1", true));
+  @Test
+  void should_detect_unsupported_vector_selector() {
+    String query = "SELECT (vector<int,3>)[0.123] FROM ks.t1";
+    QueryInspector inspector = new QueryInspector(query);
+    assertThat(inspector.hasUnsupportedSelectors()).isTrue();
   }
 
   @ParameterizedTest

--- a/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
+++ b/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/schema/QueryInspectorTest.java
@@ -64,6 +64,8 @@ class QueryInspectorTest {
   private static final CQLWord T_1 = CQLWord.fromInternal("t1");
   private static final CQLWord T_2 = CQLWord.fromInternal("t2");
 
+  private static final CQLWord VECTOR = CQLWord.fromInternal("vector");
+
   private static final CQLLiteral _16 = new CQLLiteral("16");
   private static final CQLLiteral _2 = new CQLLiteral("2");
   private static final CQLLiteral _3 = new CQLLiteral("3");
@@ -226,6 +228,26 @@ class QueryInspectorTest {
         .containsEntry(PK, PK)
         .containsEntry(CC, CC)
         .containsEntry(V, FUNC_NOW);
+  }
+
+  @Test
+  void should_detect_named_vector_insert() {
+    QueryInspector inspector =
+        new QueryInspector("INSERT INTO ks.table1 (pk, cc, vector) VALUES (:pk, :cc, :vector)");
+    assertThat(inspector.getAssignments())
+        .containsEntry(PK, PK)
+        .containsEntry(CC, CC)
+        .containsEntry(VECTOR, VECTOR);
+  }
+
+  @Test
+  void should_detect_positional_vector_insert() {
+    QueryInspector inspector =
+        new QueryInspector("INSERT INTO ks.table1 (pk, cc, vector) VALUES (?,?,?)");
+    assertThat(inspector.getAssignments())
+        .containsEntry(PK, PK)
+        .containsEntry(CC, CC)
+        .containsEntry(VECTOR, VECTOR);
   }
 
   @Test

